# Date Night Movie

date of submission: 10/03/2022

#### Grading:


- Code: 90 pts
- Markdown Documentation: 10 pts


In this assignment we are going to use pandas to figure out - What's the best **date-night movie**?

This assignment is going to use
- Joining
- Groupby
- Sorting


In [1]:
import os
import pandas as pd

##### Read in the movie data: `pd.read_table`

In [2]:
def get_movie_data():
    
    unames = ['user_id','gender','age','occupation','zip']
    users = pd.read_table(os.path.join('../data','users.dat'), 
                          sep='::', header=None, names=unames, encoding='ascii')
    
    rnames = ['user_id', 'movie_id', 'rating', 'timestamp']
    ratings = pd.read_table(os.path.join('../data', 'ratings.dat'), 
                            sep='::', header=None, names=rnames, encoding='ascii')
    
    mnames = ['movie_id', 'title','genres']
    movies = pd.read_table(os.path.join('../data', 'movies.dat'), 
                           sep='::', header=None, names=mnames, encoding='ISO-8859-1')

    return users, ratings, movies

In [3]:
users, ratings, movies = get_movie_data()

/var/folders/x4/qfzg34y540x2pv0n2m4bfkkr0000gq/T/ipykernel_90304/4026063023.py:4: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  users = pd.read_table(os.path.join('../data','users.dat'),
/var/folders/x4/qfzg34y540x2pv0n2m4bfkkr0000gq/T/ipykernel_90304/4026063023.py:8: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  ratings = pd.read_table(os.path.join('../data', 'ratings.dat'),
/var/folders/x4/qfzg34y540x2pv0n2m4bfkkr0000gq/T/ipykernel_90304/4026063023.py:12: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from 

In [4]:
users.head()

user_id gender  age  occupation    zip
0        1      F    1          10  48067
1        2      M   56          16  70072
2        3      M   25          15  55117
3        4      M   45           7  02460
4        5      M   25          20  55455

In [5]:
ratings.head()

user_id  movie_id  rating  timestamp
0        1      1193       5  978300760
1        1       661       3  978302109
2        1       914       3  978301968
3        1      3408       4  978300275
4        1      2355       5  978824291

In [6]:
movies.head()

movie_id                               title                        genres
0         1                    Toy Story (1995)   Animation|Children's|Comedy
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy
2         3             Grumpier Old Men (1995)                Comedy|Romance
3         4            Waiting to Exhale (1995)                  Comedy|Drama
4         5  Father of the Bride Part II (1995)                        Comedy

##### Clean up the `movies`

- Get the `year`
- Shorten the `title`


In [7]:
tmp = movies.title.str.extract('(.*) \(([0-9]+)\)')
tmp.apply(lambda x:x[0] if len(x) > 0 else None)
tmp.apply(lambda x: x[0][:40] if len(x) > 0 else None)

0    Toy Story
1         1995
dtype: object

In [8]:
movies['year'] = tmp[1]
movies['short_title'] = tmp[0]

In [9]:
movies.head()

movie_id                               title                        genres  \
0         1                    Toy Story (1995)   Animation|Children's|Comedy   
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy   
2         3             Grumpier Old Men (1995)                Comedy|Romance   
3         4            Waiting to Exhale (1995)                  Comedy|Drama   
4         5  Father of the Bride Part II (1995)                        Comedy   

   year                  short_title  
0  1995                    Toy Story  
1  1995                      Jumanji  
2  1995             Grumpier Old Men  
3  1995            Waiting to Exhale  
4  1995  Father of the Bride Part II

##### Join the tables with `pd.merge` (20 pts)

In [10]:
#Combine users and ratings table together such that each user has their own rating next to them
user_rating = pd.merge(users, ratings, on='user_id', how='outer')

#We combine the new 
finaltable = pd.merge(user_rating,movies, on='movie_id', how='left')

print("User table size: ",users.shape)
print("Ratings table size: ",ratings.shape)
print("Movies table size: ",movies.shape)
print("Users and Ratings table size: ",user_rating.shape)
print("Final merged table size between users, ratings, and movies: ",finaltable.shape) 
print("\n\nFinal Table: ")
finaltable

User table size:  (6040, 5)
Ratings table size:  (1000209, 4)
Movies table size:  (3883, 5)
Users and Ratings table size:  (1000209, 8)
Final merged table size between users, ratings, and movies:  (1000209, 12)


Final Table: 


user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0              1      F    1          10  48067      1193       5  978300760   
1              1      F    1          10  48067       661       3  978302109   
2              1      F    1          10  48067       914       3  978301968   
3              1      F    1          10  48067      3408       4  978300275   
4              1      F    1          10  48067      2355       5  978824291   
...          ...    ...  ...         ...    ...       ...     ...        ...   
1000204     6040      M   25           6  11106      1091       1  956716541   
1000205     6040      M   25           6  11106      1094       5  956704887   
1000206     6040      M   25           6  11106       562       5  956704746   
1000207     6040      M   25           6  11106      1096       4  956715648   
1000208     6040      M   25           6  11106      1097       4  956715569   

                                          title  \
0        One Flew Over the Cuckoo's Nest (1975)   
1              James and the Giant Peach (1996)   
2                           My Fair Lady (1964)   
3                        Erin Brockovich (2000)   
4                          Bug's Life, A (1998)   
...                                         ...   
1000204              Weekend at Bernie's (1989)   
1000205                 Crying Game, The (1992)   
1000206         Welcome to the Dollhouse (1995)   
1000207                  Sophie's Choice (1982)   
1000208       E.T. the Extra-Terrestrial (1982)   

                                  genres  year  \
0                                  Drama  1975   
1           Animation|Children's|Musical  1996   
2                        Musical|Romance  1964   
3                                  Drama  2000   
4            Animation|Children's|Comedy  1998   
...                                  ...   ...   
1000204                           Comedy  1989   
1000205                Drama|Romance|War  1992   
1000206                     Comedy|Drama  1995   
1000207                            Drama  1982   
1000208  Children's|Drama|Fantasy|Sci-Fi  1982   

                             short_title  
0        One Flew Over the Cuckoo's Nest  
1              James and the Giant Peach  
2                           My Fair Lady  
3                        Erin Brockovich  
4                          Bug's Life, A  
...                                  ...  
1000204              Weekend at Bernie's  
1000205                 Crying Game, The  
1000206         Welcome to the Dollhouse  
1000207                  Sophie's Choice  
1000208       E.T. the Extra-Terrestrial  

[1000209 rows x 12 columns]

##### What's the highest rated movie? (20 pts))

### We first take the final table and create a new table with only those rows that have rating of 5.

In [11]:
five_ratings = finaltable.loc[(finaltable['rating'] == 5)]
five_ratings

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0              1      F    1          10  48067      1193       5  978300760   
4              1      F    1          10  48067      2355       5  978824291   
6              1      F    1          10  48067      1287       5  978302039   
7              1      F    1          10  48067      2804       5  978300719   
10             1      F    1          10  48067       595       5  978824268   
...          ...    ...  ...         ...    ...       ...     ...        ...   
1000194     6040      M   25           6  11106      1077       5  964828799   
1000199     6040      M   25           6  11106      2022       5  956716207   
1000200     6040      M   25           6  11106      2028       5  956704519   
1000205     6040      M   25           6  11106      1094       5  956704887   
1000206     6040      M   25           6  11106       562       5  956704746   

                                          title                        genres  \
0        One Flew Over the Cuckoo's Nest (1975)                         Drama   
4                          Bug's Life, A (1998)   Animation|Children's|Comedy   
6                                Ben-Hur (1959)        Action|Adventure|Drama   
7                     Christmas Story, A (1983)                  Comedy|Drama   
10                  Beauty and the Beast (1991)  Animation|Children's|Musical   
...                                         ...                           ...   
1000194                          Sleeper (1973)                 Comedy|Sci-Fi   
1000199   Last Temptation of Christ, The (1988)                         Drama   
1000200              Saving Private Ryan (1998)              Action|Drama|War   
1000205                 Crying Game, The (1992)             Drama|Romance|War   
1000206         Welcome to the Dollhouse (1995)                  Comedy|Drama   

         year                      short_title  
0        1975  One Flew Over the Cuckoo's Nest  
4        1998                    Bug's Life, A  
6        1959                          Ben-Hur  
7        1983               Christmas Story, A  
10       1991             Beauty and the Beast  
...       ...                              ...  
1000194  1973                          Sleeper  
1000199  1988   Last Temptation of Christ, The  
1000200  1998              Saving Private Ryan  
1000205  1992                 Crying Game, The  
1000206  1995         Welcome to the Dollhouse  

[226310 rows x 12 columns]

### Then we take the movie id that has the highest number of frequency in the movie_id column. In other words we pick the movie id that is rated 5 by the majority of the users. 

In [12]:
print(five_ratings['movie_id'].value_counts())
common_movie = five_ratings['movie_id'].value_counts().idxmax()
print("\nThe movie_id with the highest frequency of 5 ratings amongst males and females is: ",common_movie)

2858    1963
260     1826
1198    1500
1196    1483
527     1475
        ... 
460        1
2758       1
1807       1
114        1
2992       1
Name: movie_id, Length: 3232, dtype: int64

The movie_id with the highest frequency of 5 ratings amongst males and females is:  2858


### The table above displays the movie id's on the left and how many users rated 5 to that movie on the right. I.E., movie id 2858 was rated 5 by 1963 users.

###### What is a good rated movie for date night? (60 pts)

- Hint - highly rated movie by 
    - both partners (might be the same gender or not),
    - based on genre preferences,
    - age group can also be combined

### Then we use that movie id to find the movie that is the highest rated amongst the people as a good movie for date night. And this movie is the best for date night because the majority of the users prefer this movie over the others.

In [13]:
print("The greatest rated movie for date night according to the people is:\n",
      movies.loc[movies['movie_id'] == common_movie, 'title'].iloc[0])

The greatest rated movie for date night according to the people is:
 American Beauty (1999)
